app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,437 @@
+      } else {
+        ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+            PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+      }
+      return false;
+
+    } else {
+      return true; // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  private boolean checkExternalStorageWritePermission() { // To access and store the zims
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+        == PackageManager.PERMISSION_DENIED) {
+
+      if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+        alertDialogShower.show(KiwixDialog.StoragePermissionRationale.INSTANCE,
+            new Function0<Unit>() {
+              @Override public Unit invoke() {
+                ActivityCompat.requestPermissions(LocalFileTransferActivity.this, new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+                    PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+                return Unit.INSTANCE;
+              }
+            });
+      } else {
+        ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+            PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+      }
+      return false;
+
+    } else {
+      return true; // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+
+    if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
+      switch (requestCode) {
+
+        case PERMISSION_REQUEST_CODE_COARSE_LOCATION: {
+          Log.e(TAG, "Location permission not granted");
+
+          showToast(this, R.string.permission_refused_location, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        case PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS: {
+          Log.e(TAG, "Storage write permission not granted");
+
+          showToast(this, R.string.permission_refused_storage, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        default: {
+          super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+          break;
+        }
+      }
+    }
+  }
+
+  private boolean isLocationServicesEnabled() {
+    LocationManager locationManager =
+        (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+    boolean gps_enabled = false;
+    boolean network_enabled = false;
+
+    try {
+      gps_enabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    try {
+      network_enabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+    } catch (Exception ex) {
```

Comment:
Don't catch Exception.
If code throws multiple Exceptions you can Or them together `catch(IllegalArgumentException | SecurityException exception)`

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/cb944bb88f3809b3baee4613e21e92972122a775#diff-e1d59ae6e15dcb502e125fb6e243d15dR340)
Does object slicing from implicit casting affect something like the stack-trace as well?

Reply:
```
This issue is not inherently unique to C++, but it does not occur naturally in most other object-oriented languages — even C++'s relatives such as D, Java, and C# — because copying of objects is not a basic operation in those languages.
```
I had no idea what you were talking about. Java is your friend and does not do this